### PR TITLE
feat: add PKCE for Oauth2 OIDC

### PR DESCRIPTION
**Pull request by anderska**
### Summary
This PR enhances the OAuth2/OIDC integration with:
- Configurable token endpoint authentication (basic, post, none)
- Optional PKCE support via code challenge method
- Explicit required group claim enforcement (non-empty claims)
- Simplified, safer scopes handling and claim parsing
- Removal of brittle token “granted scope” checks in favor of claim-based validation

### Changes compared to main

- OAuth2 configuration (src/phoenix/config.py)
  - Fixed default scopes to `openid email profile`.
  - New env: `PHOENIX_OAUTH2_{IDP}_REQUIRED_GROUPS`
    - Space-separated claim names you want the IdP to return (e.g., `groups roles custom_claim`).
    - Appended raw to the requested scope string.
    - Parsed into `required_groups` for server-side enforcement.
  - New env: `PHOENIX_OAUTH2_{IDP}_TOKEN_ENDPOINT_AUTH_METHOD`
    - Allowed: `client_secret_basic` (default), `client_secret_post`, `none`.
    - Propagated to the OAuth2 client for token exchange.
  - Existing env: `PHOENIX_OAUTH2_{IDP}_CODE_CHALLENGE_METHOD`
    - Allowed: `S256` (recommended) or `plain`.
    - Enables PKCE (sends code_challenge + code_verifier).
  - Updated inline docs and env discovery to include these keys.

- OAuth2 client (src/phoenix/server/oauth2.py)
  - Passes through:
    - `scope` from config,
    - `required_groups` (for downstream logic),
    - `token_endpoint_auth_method` when set,
    - `code_challenge_method` when set.
  - Uses built-in `AsyncHttpxOAuth2Client` with the above kwargs.

- Router and auth flow (src/phoenix/server/api/routers/oauth2.py)
  - PKCE wired end-to-end: stores `code_verifier` in a cookie and sends it on token exchange.
  - `parse_user_info`: builds `claims` containing only non-empty values (filters out `None`, empty strings/containers).
  - Authorization now validates required claim keys (from `required_groups`) exist and are non-empty; otherwise raises `SignInNotAllowed`.
  - Removed strict “requested ⊆ granted” token scope check in favor of claim-based validation.
  - Removed debug logging.

### New/updated environment variables

- `PHOENIX_OAUTH2_{IDP}_REQUIRED_GROUPS`
  - Space-separated list of claim names the IdP should include in the ID token (e.g., `groups roles team`).
  - Appended to requested scopes as-is.
  - Sign-in requires each listed claim to exist and be non-empty.

- `PHOENIX_OAUTH2_{IDP}_TOKEN_ENDPOINT_AUTH_METHOD`
  - Token endpoint client auth mode.
  - Allowed: `client_secret_basic` (default), `client_secret_post`, `none`.

- `PHOENIX_OAUTH2_{IDP}_CODE_CHALLENGE_METHOD`
  - Enables PKCE when set.
  - Allowed: `S256` (recommended) or `plain`.

### Examples

```bash
# Ask IdP to include these claim keys and require them be non-empty
export PHOENIX_OAUTH2_GOOGLE_REQUIRED_GROUPS="groups roles team"

# Use client_secret_post to send client credentials in the token request body
export PHOENIX_OAUTH2_GOOGLE_TOKEN_ENDPOINT_AUTH_METHOD="client_secret_post"

# Enable PKCE with S256
export PHOENIX_OAUTH2_GOOGLE_CODE_CHALLENGE_METHOD="S256"
```

---

> [!NOTE]
> Adds PKCE, token endpoint auth method config, JMESPath-based group extraction/enforcement, and safer claim handling to OAuth2/OIDC with comprehensive tests.
> 
> - **OAuth2/OIDC**:
>   - **PKCE support**:
>     - Set/validate `code_verifier` via new cookie `PHOENIX_OAUTH2_CODE_VERIFIER_COOKIE_NAME`; send during token exchange; cleanup on success/error.
>   - **Token exchange**:
>     - Configurable `token_endpoint_auth_method` (`client_secret_basic|client_secret_post|none`); HTTPS URL validation; baseline scopes `openid email profile`.
>   - **Claim handling**:
>     - Parse/validate ID token claims robustly (`InvalidUserInfo`, stricter email checks); fetch and merge `userinfo` claims when needed.
>   - **Group-based access control**:
>     - New config: `GROUPS_ATTRIBUTE_PATH` (JMESPath) and `ALLOWED_GROUPS`; precompiled JMESPath, normalization, and enforcement in `OAuth2Client.validate_access`.
> - **Config/Client**:
>   - `OAuth2ClientConfig` extended with `use_pkce`, `token_endpoint_auth_method`, `scopes`, `groups_attribute_path`, `allowed_groups`; improved env scanning/validation.
>   - `OAuth2Client` extended to honor scopes, PKCE (`S256`), auth method, and group rules; adds `has_sufficient_claims`.
> - **API Router** (`server/api/routers/oauth2.py`):
>   - Validates state/nonce, optional PKCE, merges userinfo, enforces access, and cleans up cookies.
> - **Deps/Types**:
>   - Add `jmespath` runtime dep and `types-jmespath` for type-checking.
> - **Tests**:
>   - Extensive unit/integration coverage for PKCE (public/confidential), cookie security, state/CSRF, group access (granted/denied), JMESPath validation, and config edge cases.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 96f1c04a902946fc39f6d58870f4a37fb9c08a53. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by axiomofjoy** [comment]
Thanks @anderska for the contribution. The team is heads down on feature work and I'm not deeply familiar with PKSE, so it might take a minute, but we'll take a look.

**Comment by RogerHYang** [comment]
Thanks for your PKCE implementation! I've made several design changes that I wanted to explain:

---

## **1. `CODE_CHALLENGE_METHOD` → `USE_PKCE` (boolean)**

**Your approach:** Exposed `CODE_CHALLENGE_METHOD` as a user-configurable string (`S256` or `plain`)

**Changed to:** Simple boolean `USE_PKCE=true` with S256 hardcoded

**Rationale:** 
- `plain` is deprecated and insecure (RFC 7636 strongly discourages it)
- S256 is the only option anyone should use in production
- Exposing the choice adds configuration complexity without real-world benefit

Also made `CLIENT_SECRET` optional when PKCE is enabled to properly support public clients.

---

## **2. `REQUIRED_GROUPS` → `GROUPS_ATTRIBUTE_PATH` + `ALLOWED_GROUPS`**

**Your approach:** Appended group names to the `scopes` parameter, checked for their presence as top-level claims

**Changed to:** Two-part system:
- `GROUPS_ATTRIBUTE_PATH`: Path expression to extract groups from any claim structure
- `ALLOWED_GROUPS`: Authorization list
- *(Variable names align with Grafana's convention for consistency)*

**Rationale:**

The original approach had two fundamental issues:

1. **Groups aren't standardized in OIDC.** Different IDPs may structure them completely differently.

2. **Appending to `scopes` doesn't control claim contents.** Scopes like `openid`, `profile`, `email` are standard OIDC, but custom values like `admin` or `developers` aren't meaningful to the IDP. What claims are returned is configured at the IDP level, not requested via scopes.

The new approach uses a flexible path expression that works with any IDP's structure without needing IDP-specific code.

---

## **3. ID Token → UserInfo Fallback Strategy**

**Added:** Conditional userinfo endpoint call with smart fallback

**How it works:**
1. Parse ID token - if groups present, use them (no extra HTTP call)
2. If groups missing, fetch from userinfo endpoint and merge
3. If userinfo fails, fall back to ID token alone

**Rationale:**

Different IDPs have different behaviors:
- **AWS Cognito, Azure AD:** Often omit groups from ID token to keep it small; require userinfo call
- **Okta, Auth0:** Configurable - can be in either place
- **Keycloak:** Usually in ID token by default

---

Let me know if you have questions about any of these decisions - happy to discuss further!

File: src/phoenix/config.py [hunk 1]
```diff
@@ -968,88 +968,257 @@ def get_env_smtp_validate_certs() -> bool:
...
+            client_secret = _get_optional("CLIENT_SECRET")
+        else:
+            # Traditional flow: client_secret is required
+            client_secret = _get_required("CLIENT_SECRET", "Client secret")
+
+        # Token endpoint auth method validation
+        token_endpoint_auth_method = None
+        if auth_method := _get_optional("TOKEN_ENDPOINT_AUTH_METHOD"):
+            auth_method = auth_method.lower()
+            allowed_methods = {"client_secret_basic", "client_secret_post", "none"}
```
Comment (axiomofjoy) [inline src/phoenix/config.py]: Let's document these/ link to the part of the OIDC spec or other resource

File: src/phoenix/config.py [hunk 2]
```diff
@@ -968,88 +968,267 @@ def get_env_smtp_validate_certs() -> bool:
...
 
 
+_OAUTH2_CONFIG_SUFFIXES = (
+    "DISPLAY_NAME",  # User-friendly name shown in login UI
+    "CLIENT_ID",  # OAuth2 client ID from your identity provider
+    "CLIENT_SECRET",  # OAuth2 client secret (required unless using PKCE)
+    "OIDC_CONFIG_URL",  # OpenID Connect discovery URL (.well-known/openid-configuration)
+    "ALLOW_SIGN_UP",  # Whether to allow new user registration (default: true)
+    "AUTO_LOGIN",  # Automatically redirect to this provider (default: false)
+    "USE_PKCE",  # Enable PKCE for public clients like mobile apps (default: false)
```
Comment (axiomofjoy) [inline src/phoenix/config.py]: PKCE is not just for public clients in my understanding.

File: src/phoenix/config.py [hunk 3]
```diff
@@ -968,88 +968,267 @@ def get_env_smtp_validate_certs() -> bool:
...
+    "TOKEN_ENDPOINT_AUTH_METHOD",  # How to authenticate at token endpoint
```
Comment (axiomofjoy) [inline src/phoenix/config.py]: What is this config?

File: src/phoenix/server/api/routers/oauth2.py [hunk 4]
```diff
@@ -173,27 +186,54 @@ async def create_tokens(
...
+        # PKCE validation: code_verifier is required when PKCE is enabled
+        if oauth2_client.use_pkce:
+            if not code_verifier:
+                logger.error(
+                    "PKCE enabled but code_verifier cookie missing for IDP %s. "
+                    "This may indicate a cookie issue, CORS misconfiguration, or "
+                    "browser compatibility problem.",
+                    idp_name,
+                )
+                return _redirect_to_login(request=request, error="auth_failed")
```
Comment (axiomofjoy) [inline src/phoenix/server/api/routers/oauth2.py:207]:
```suggestion
                return _redirect_to_login(request=request, error="Login failed.")
```

This error is displayed to users in the UI.

File: src/phoenix/server/api/routers/oauth2.py [hunk 5]
```diff
@@ -173,27 +186,54 @@ async def create_tokens(
...
-    user_info = await oauth2_client.parse_id_token(token_data, nonce=stored_nonce)
+
+    id_token_claims = await oauth2_client.parse_id_token(token_data, nonce=stored_nonce)
+
+    if oauth2_client.has_sufficient_claims(id_token_claims):
+        user_claims = id_token_claims
+    else:
+        user_claims = await _fetch_and_merge_userinfo_claims(
+            oauth2_client, token_data, id_token_claims
+        )
```
Comment (axiomofjoy) [inline src/phoenix/server/api/routers/oauth2.py:225]: I'd lean on the side of not including this extra logic until we've seen that it is necesssary for a specific IDP. I just wouldn't take ChatGPT's word for it.

```suggestion
    if oauth2_client.has_sufficient_claims(id_token_claims):
        user_claims = id_token_claims
    else:
        user_claims = await _fetch_and_merge_userinfo_claims(
            oauth2_client, token_data, id_token_claims
        )
```

File: src/phoenix/server/api/routers/oauth2.py [hunk 6]
```diff
@@ -173,27 +186,54 @@ async def create_tokens(
...
+
     try:
-        user_info = _parse_user_info(user_info)
-    except MissingEmailScope as e:
-        logger.error("Missing email scope for IDP %s: %s", idp_name, e)
+        user_info = _parse_user_info(user_claims)
+    except (MissingEmailScope, InvalidUserInfo) as e:
+        logger.error("Error parsing user info for IDP %s: %s", idp_name, e)
         return _redirect_to_login(request=request, error="missing_email_scope")
```
Comment (axiomofjoy) [inline src/phoenix/server/api/routers/oauth2.py:231]: This error message is displayed in the UI.

File: src/phoenix/server/api/routers/oauth2.py [hunk 7]
```diff
@@ -173,27 +186,54 @@ async def create_tokens(
...
 
+    try:
+        oauth2_client.validate_access(user_info.claims)
+    except PermissionError as e:
+        logger.error("Access validation failed for IDP %s: %s", idp_name, e)
+        return _redirect_to_login(request=request, error="auth_failed")
```
Comment (axiomofjoy) [inline src/phoenix/server/api/routers/oauth2.py:237]: Error message displayed in the UI.

**Review by axiomofjoy — APPROVED** [review]

File: src/phoenix/server/api/routers/oauth2.py (hunk 5 above)
Comment (RogerHYang) [reply src/phoenix/server/api/routers/oauth2.py:225]: The basis for this logic is in the [specification](https://openid.net/specs/openid-connect-core-1_0.html) itself, so there’s no need to treat it as optional or speculative.

5.1 Standard Claims
This specification defines a set of standard Claims. They can be requested to be returned either in the UserInfo Response, per [Section 5.3.2](https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse), or in the ID Token, per [Section 2](https://openid.net/specs/openid-connect-core-1_0.html#IDToken).

5.1.2.  Additional Claims
[...] other Claims MAY be used in conjunction with the standard Claims.

File: src/phoenix/server/api/routers/oauth2.py (hunk 6 above)
Comment (RogerHYang) [reply src/phoenix/server/api/routers/oauth2.py:231]: these are error codes and not displayed in the UI

File: src/phoenix/config.py (hunk 1 above)
Comment (RogerHYang) [reply src/phoenix/config.py]: these are documented in a consolidated docstring for `get_env_oauth2_settings`

File: src/phoenix/config.py (hunk 3 above)
Comment (RogerHYang) [reply src/phoenix/config.py]: this is documented in the consolidated docstring for 'get_env_oauth2_settings'
